src/machines/java/com/enderio/machines/common/recipe/FermentingRecipe.java

```diff
@@ -0,0 +1,202 @@
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.tags.ItemTags;
+import net.minecraft.tags.TagKey;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.RecipeSerializer;
+import net.minecraft.world.item.crafting.RecipeType;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.material.Fluid;
+import net.neoforged.neoforge.fluids.FluidStack;
+import net.neoforged.neoforge.items.IItemHandlerModifiable;
+import net.neoforged.neoforge.items.wrapper.RecipeWrapper;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.Objects;
+
+public class FermentingRecipe implements MachineRecipe<FermentingRecipe.Container> {
+
+    private final FluidIngredient inputFluid;
+    private final int inputFluidAmount;
+    private final TagKey<Item> leftReagent;
+    private final TagKey<Item> rightReagent;
+    private final double outputModifier;
+    private final Fluid outputFluid;
+    private final int ticks;
+
+    public FermentingRecipe(FluidIngredient inputFluid, int inputFluidAmount, TagKey<Item> leftReagent, TagKey<Item> rightReagent, Fluid outputFluid,
+        double outputModifier, int ticks) {
+        this.inputFluid = inputFluid;
+        this.inputFluidAmount = inputFluidAmount;
+        this.leftReagent = leftReagent;
+        this.rightReagent = rightReagent;
+        this.outputModifier = outputModifier;
+        this.outputFluid = outputFluid;
+        this.ticks = ticks;
+    }
+
+    @Override
+    public int getBaseEnergyCost() {
+        return 0;
+    }
+
+    @Override
+    public List<OutputStack> craft(Container container, RegistryAccess registryAccess) {
+
+        double totalModifier = outputModifier;
+        totalModifier *= getReagentModifier(container.getItem(0), leftReagent);
+        totalModifier *= getReagentModifier(container.getItem(1), rightReagent);
+
+        return List.of(OutputStack.of(new FluidStack(outputFluid, (int) (inputFluidAmount * totalModifier))));
+    }
+
+    @Override
+    public List<OutputStack> getResultStacks(RegistryAccess registryAccess) {
+        return List.of(OutputStack.EMPTY);
+    }
+
+    @Override
+    public boolean matches(Container container, Level level) {
+        FluidStack inputTank = container.getInputTank().getFluid();
+        if (!inputFluid.test(inputTank.getFluid()) || inputTank.getAmount() < inputFluidAmount) {
+            return false;
+        }
+        if (!container.getItem(0).is(leftReagent)) {
+            return false;
+        }
+        if (!container.getItem(1).is(rightReagent)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    public double getReagentModifier(ItemStack stack, TagKey<Item> reagent) {
+        var map = stack.getItemHolder().getData(VatReagent.DATA_MAP);
+        if (map != null) {
+            return map.getOrDefault(reagent, 1D);
+        }
+        return 1;
```

Comment:
Hmm, should 1 be the default? This code path shouldn't ever be run, but would 0 not be better?

Reply:
This piece of code is run when the output is generated (not when the recipe match happens) and so returning 0 can cause issues like having 0 output even after consuming inputs which is why it returns 1 as in doesn't affect output at all. 
 Actually, I think I should add a check in recipe match as well, just in case.